site-src/geps/gep-1364.md

```diff
@@ -0,0 +1,328 @@
+set to `true` means that sufficient capacity exists on underlying infrastructure
+for the Gateway to be provisioned. If that capacity does not exist, then the
+Gateway cannot be reconciled successfully, and so fails to attach to the
+owning GatewayClass.
+
+Note that some classes of inter-resource reference failure do _not_ cause a resource
+to become unattached (that is, to have the `Attached` condition set to `status: false`).
+* Nonexistent Service backends - if the backend does not exist on a HTTPRoute that
+is otherwise okay, then the data plane must generate 500s for traffic that matches
+that HTTPRoute.
+* HTTPRoutes with *all* backends in other namespaces, but not permitted by ReferenceGrants.
+In this case, the "nonexistent service backends" rules apply, and 500s must be
+generated.
+
+For ReferenceGrant or not-designed-yet Policy resources, `Attached` means that:
+* the resource has a correctly-defined set of resources that it applies to
+* the resource has a syntactically and semantically valid `spec`
+
+Note that having a correctly-defined set of resources that is empty does not make
+these resources unattached.
+
+Note that for other Route types that don't have a clear mechanism like HTTP does
+for indicating a server failure (like the HTTP code 500 does), not having existing
+backends may not produce any configuration in the data plane, and so may cause
+the resource to fail to attach. (An example here could be a TCP Route with
+no backends, we need to decide if that means that a port should be opened that
+actively closes connections, or if no port should be opened.)
+
+
+#### Ready
+
+Currently, the `Ready` condition text for Gateway says:
+```go
+	// This condition is true when the Gateway is expected to be able
+	// to serve traffic. Note that this does not indicate that the
+	// Gateway configuration is current or even complete (e.g. the
+	// controller may still not have reconciled the latest version,
+	// or some parts of the configuration could be missing).
+```
+
+This is pretty unclear - how can the Gateway serve traffic if config is missing?
+In the past, we've been asked to have a Condition that only flips to `true` when
+*all* required configuration is present.
+
+For many implementations (certainly for Envoy-based ones), getting this information
+correctly and avoiding races on applying it is surprisingly difficult. For this
+reason, this GEP proposes we give the `Ready` condition a short window
+
+This condition needs a few changes:
+
+* Each resource should have a `Ready` condition
+* The `Ready` condition should be clear as to *when* the traffic is able to pass
+  the implementation. Since it's _extremely_ difficult to _guarantee_ this, we're
+  stuck with some sort of "will be ready in no longer than x seconds" standard
+  (which we'll also need to write conformance tests for, probably including a 
+  stress test of adding a *lot* of records as well.) As an initial ballpark,
+  I propose 5 seconds.
```

Comment:
This probably won't work for cloud providers (at least GKE is slower, not sure about others).

I am also not sure it provides value. What do we lose if we don't define anything here?

If we consider something like Knative, they certainly aren't going to be able to just wait X seconds; I'd expect the same with other users

Reply:
And if we can't provide any guarantees about Readiness, then it seems like Ready becomes no different than Attached...

Reply:
We certainly can mandate `Ready` to be an _extended_ condition, and say "If you supply a Ready condition, it has to work like this". Maybe doing that is a better place to start with `Ready`?

Reply:
Do we have a feel for which implementations can and cannot support `Ready` properly?

Reply:
Agree that cloud providers will likely have trouble implementing a `Ready` condition. Could we introduce a `Reconciled` condition that would be more broadly implementable?

Reply:
+1 any distributed (e.g. geo distributed) implementation has a hard time with this (think: config was pushed to 95% of the sites around the world, is it ready?)